test/cpp_test_suite/new_tests/cxx_cmd_types.cpp

```diff
@@ -879,24 +879,88 @@ class CmdTypesTestSuite: public CxxTest::TestSuite
     {
         for (size_t i = 0; i < loop; i++)
         {
-            DeviceData din, dout;
             DevicePipeBlob in{"TestPipeCmd"};
-            vector<string> names{"level0"};
+            DeviceData din, dout;
+            DeviceData din2, dout2;
+
+            vector <string> names{"long", "double", "string", "double array"};
             in.set_data_elt_names(names);
-            DevLong data = 123;
-            in << data;
-            din.insert(in);
-            TS_ASSERT_THROWS_NOTHING(dout = device->command_inout("IOPipeBlob", din));
-            DevicePipeBlob received{};
-            dout.extract(&received);
+            DevLong long_data = 123;
+            TS_ASSERT_THROWS_NOTHING(in["long"] << long_data);
+            DevDouble double_data = 3.14;
+            TS_ASSERT_THROWS_NOTHING(in["double"] << double_data);
+            string data_str = "String value";
+            TS_ASSERT_THROWS_NOTHING(in["string"] << data_str);
+            vector<double> double_array = {0.1, -0.02, 0.003, -0.0004};
+            TS_ASSERT_THROWS_NOTHING(in["double array"] << double_array);
+            TS_ASSERT_THROWS_NOTHING(din << in);
 
```

Comment:
I would first stay with more intuitive approach and only if necessary (users report performance issues + benchmark test) will try to optimize it. 

In fact what is need here is builder pattern: first user modifies DevPipeBlobBuilder with the data then freezes it (acquiring immutable DevicePipeBlob) and pass it to command_inout OR we can perform defensive copying inside command_inout